### DISCUSSION TITLE
navigation ref API inconsistency with navigation provided through injected prop or useNavigation()

### DISCUSSION THREAD

**Discussion by jaredsavage** [react-navigation/react-navigation#11263] [Q&A] [answered]
Hi there!

I recently switched over to react-navigation for a medium-sized project, and didn't realize initially that `.pop()`,`.replace()` methods are not defined on navigation when it's accessed through createNavigationContainerRef, and only when accessed via injected prop or useNavigation.

We're using Sagas extensively for navigation control, and would prefer to have a single API for navigation if possible.

I understand that the way to access these is to use dispatch:
```
navigationRef.dispatch(StackActions.replace('xyz'))
```

What I'm wondering is, why not include `.pop()`, `.popToTop()`, `.push()`, `.replace()` in the set of functions available on navigationRef, and map to this internally? My understanding is that this would allow navigation to have a consistent API everywhere it's used. If a Stack.navigator isn't defined / no matching route exists, an console error would do the trick to inform the dev they missed a step?

My view is definitely limited by my use case, which has little nesting, and a single Stack.Navigator. That said, it seems like this convenience and consistency would be more helpful than confusing? I'd like to understand what I'm missing.

**Comment by satya164** [comment, marked as answer]
When a navigator renders, it adds the necessary methods to the navigation object. e.g. you won't see `push`, `pop` etc. in a tab navigator unless it's inside a stack. Similarly, the ref doesn't have these methods since it's not inside a stack navigator.

Sure it could contain these methods but then they won't be type-safe and may work or fail based on what navigators are rendered - which makes them flaky. You can also export such utilities yourself if you need a nicer API like shown in the docs.

Also, keep in mind that the ref is also an escape hatch, you should avoid using it unless absolutely necessary.

**Reply by jaredsavage** [reply]
Thanks for the quick reply. That makes sense. I did go the route of making a little helper to wrap and include these.

Our use case has a pretty simple and static route set at the moment, so having a best effort works well for us. 

I'm wondering, for working within Sagas, in your opinion what would be the best option to use?

For navigation/stack actions `(navigate, pop, reset, replace, etc)`, `dispatch` works, but for when we need to check the current route, route params, or get a list of available routes, is the ref the best bet?
